File: json_bench_python/json_bench_python/core/commands/map_command.py
"""map_values-style command implementation."""

from __future__ import annotations

import math
import re
from dataclasses import dataclass
from decimal import Decimal, InvalidOperation, ROUND_HALF_UP
from typing import Any, Sequence

from .filter_command import FilterCommand, json_type_name


ARITHMETIC_EXPR = re.compile(r"^\s*\.?\s*([+\-*/^])\s*([+-]?(?:\d+(?:\.\d+)?|\.\d+))\s*$")
DIVISION_SCALE = Decimal("1." + ("0" * 16))


@dataclass(frozen=True)
class ArithmeticOp:
    operator: str
    operand: Decimal


class MapCommand:
    """Applies arithmetic transforms or filter expressions per value/element."""

    def __init__(self) -> None:
        self._filter_command = FilterCommand()

    def name(self) -> str:
        return "map"

    def execute(self, input_data: Any, args: Sequence[str] | None) -> Any:
        if input_data is None:
            return None

        filter_expr = "." if not args else args[0]

        if isinstance(input_data, dict):
            return {key: self._apply_expression(value, filter_expr) for key, value in input_data.items()}

        if isinstance(input_data, list):
            return [self._apply_expression(item, filter_expr) for item in input_data]

        raise ValueError(f"map_values expects object or array input, got: {json_type_name(input_data)}")

    def _apply_expression(self, node: Any, filter_expr: str) -> Any:
        arithmetic_op = self._parse_arithmetic(filter_expr)
        if arithmetic_op is not None:
            return self._apply_arithmetic_recursively(node, arithmetic_op)

        return self._filter_command.execute(node, [filter_expr])

    def _parse_arithmetic(self, expr: str) -> ArithmeticOp | None:
        matcher = ARITHMETIC_EXPR.match(expr)
        if not matcher:
            return None

        operator = matcher.group(1)
        operand = Decimal(matcher.group(2))

        if operator == "/" and operand == 0:
            raise ValueError(f"Division by zero is not allowed in map expression: {expr}")

        return ArithmeticOp(operator, operand)

    def _apply_arithmetic_recursively(self, node: Any, op: ArithmeticOp) -> Any:
        if node is None:
            return None

        if isinstance(node, dict):
            return {key: self._apply_arithmetic_recursively(value, op) for key, value in node.items()}

        if isinstance(node, list):
            return [self._apply_arithmetic_recursively(item, op) for item in node]

        if not is_json_number(node):
            return node

        value = Decimal(str(node))
        calculated = self._apply_operation(value, op)
        return as_json_number(calculated)

    def _apply_operation(self, value: Decimal, op: ArithmeticOp) -> Decimal:
        if op.operator == "+":
            return value + op.operand
        if op.operator == "-":
            return value - op.operand
        if op.operator == "*":
            return value * op.operand
        if op.operator == "/":
            return (value / op.operand).quantize(DIVISION_SCALE, rounding=ROUND_HALF_UP)
        if op.operator == "^":
            return pow_decimal(value, op.operand)

        raise ValueError(f"Unsupported operator: {op.operator}")


def is_json_number(value: Any) -> bool:
    return isinstance(value, (int, float, Decimal)) and not isinstance(value, bool)


def as_json_number(value: Decimal) -> int | float:
    normalized = value.normalize()

    if normalized == normalized.to_integral_value():
        return int(normalized)

    return float(normalized)


def pow_decimal(base: Decimal, exponent: Decimal) -> Decimal:
    normalized_exponent = exponent.normalize()

    if normalized_exponent == normalized_exponent.to_integral_value():
        int_exponent = int(normalized_exponent)

        if int_exponent >= 0:
            return base ** int_exponent

        if base == 0:
            raise ValueError("0 cannot be raised to a negative power.")

        result = Decimal(1) / (base ** abs(int_exponent))
        return result.quantize(DIVISION_SCALE, rounding=ROUND_HALF_UP)

    if base < 0:
        raise ValueError("Negative base with non-integer exponent is not a real number.")

    result = math.pow(float(base), float(exponent))
    if math.isnan(result) or math.isinf(result):
        raise ValueError("Power operation produced a non-finite result.")

    try:
        return Decimal(str(result))
    except InvalidOperation as exc:
        raise ValueError("Power operation produced an invalid numeric result.") from exc

File: json_bench_java/src/main/java/benchmark/core/commands/MapCommand.java
package benchmark.core.commands;

import benchmark.core.Command;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.DecimalNode;
import com.fasterxml.jackson.databind.node.LongNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class MapCommand implements Command
{
    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final Pattern ARITHMETIC_EXPR = Pattern.compile("^\\s*\\.?\\s*([+\\-*/^])\\s*([+-]?(?:\\d+(?:\\.\\d+)?|\\.\\d+))\\s*$");

    private final FilterCommand filterCommand = new FilterCommand();

    @Override
    public String name() { return "map"; }

    @Override
    public JsonNode execute(JsonNode input, List<String> args)
    {
        if (input == null)
            return com.fasterxml.jackson.databind.node.NullNode.instance;

        String filterExpr = (args == null || args.isEmpty()) ? "." : args.getFirst();

        if (input.isObject())
        {
            ObjectNode mapped = MAPPER.createObjectNode();
            input.fields().forEachRemaining(entry -> mapped.set(
                    entry.getKey(),
                    applyExpression(entry.getValue(), filterExpr)
            ));
            return mapped;
        }

        if (input.isArray())
        {
            ArrayNode mapped = MAPPER.createArrayNode();
            for (JsonNode item : input)
            {
                mapped.add(applyExpression(item, filterExpr));
            }
            return mapped;
        }

        throw new IllegalArgumentException("map_values expects object or array input, got: " + input.getNodeType());
    }

    private JsonNode applyExpression(JsonNode node, String filterExpr)
    {
        ArithmeticOp arithmeticOp = parseArithmetic(filterExpr);

        if (arithmeticOp != null)
        {
            return applyArithmeticRecursively(node, arithmeticOp);
        }

        return filterCommand.execute(node, List.of(filterExpr));
    }

    private ArithmeticOp parseArithmetic(String expr)
    {
        Matcher matcher = ARITHMETIC_EXPR.matcher(expr);

        if (!matcher.matches())
            return null;

        String operator = matcher.group(1);
        BigDecimal operand = new BigDecimal(matcher.group(2));

        if (operator.equals("/") && BigDecimal.ZERO.compareTo(operand) == 0)
            throw new IllegalArgumentException("Division by zero is not allowed in map expression: " + expr);

        return new ArithmeticOp(operator.charAt(0), operand);
    }

    private JsonNode applyArithmeticRecursively(JsonNode node, ArithmeticOp op)
    {
        if (node == null || node.isNull())
            return com.fasterxml.jackson.databind.node.NullNode.instance;

        if (node.isObject())
        {
            ObjectNode result = MAPPER.createObjectNode();
            node.fields().forEachRemaining(entry -> result.set(entry.getKey(), applyArithmeticRecursively(entry.getValue(), op)));
            return result;
        }

        if (node.isArray())
        {
            ArrayNode result = MAPPER.createArrayNode();
            for (JsonNode item : node)
            {
                result.add(applyArithmeticRecursively(item, op));
            }
            return result;
        }

        if (!node.isNumber())
            return node;

        BigDecimal value = node.decimalValue();
        BigDecimal calculated = switch (op.operator)
        {
            case '+' -> value.add(op.operand);
            case '-' -> value.subtract(op.operand);
            case '*' -> value.multiply(op.operand);
            case '/' -> value.divide(op.operand, 16, java.math.RoundingMode.HALF_UP);
            case '^' -> pow(value, op.operand);
            default -> throw new IllegalArgumentException("Unsupported operator: " + op.operator);
        };

        return asJsonNumber(calculated);
    }

    private JsonNode asJsonNumber(BigDecimal value)
    {
        BigDecimal normalized = value.stripTrailingZeros();

        if (normalized.scale() <= 0)
        {
            try
            {
                return LongNode.valueOf(normalized.longValueExact());
            } catch (ArithmeticException ignored) {}
        }

        return DecimalNode.valueOf(normalized);
    }

    private BigDecimal pow(BigDecimal base, BigDecimal exponent)
    {
        BigDecimal normalizedExponent = exponent.stripTrailingZeros();

        if (normalizedExponent.scale() <= 0)
        {
            try
            {
                int intExponent = normalizedExponent.intValueExact();

                if (intExponent >= 0)
                    return base.pow(intExponent);

                if (BigDecimal.ZERO.compareTo(base) == 0)
                    throw new IllegalArgumentException("0 cannot be raised to a negative power.");

                return BigDecimal.ONE.divide(base.pow(-intExponent), 16, RoundingMode.HALF_UP);
            } catch (ArithmeticException ignored) {}
        }

        if (base.compareTo(BigDecimal.ZERO) < 0)
            throw new IllegalArgumentException("Negative base with non-integer exponent is not a real number.");

        double result = Math.pow(base.doubleValue(), exponent.doubleValue());
        if (Double.isNaN(result) || Double.isInfinite(result))
            throw new IllegalArgumentException("Power operation produced a non-finite result.");

        return BigDecimal.valueOf(result);
    }

    private record ArithmeticOp(char operator, BigDecimal operand) {
    }
}
